datafusion/src/datasource/protocol_registry.rs

```diff
@@ -0,0 +1,83 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::datasource::local::LocalFSHandler;
+use crate::error::Result;
+use parquet::file::reader::ChunkReader;
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+
+pub trait ProtocolHandler<R>: Sync + Send {
```

Comment:
I wonder if calling interface `ObjectStore` might be more standard -- it is basically an abstraction for something that maps string keys to sources of "objects" (aka bytes). 

I also suggest adding some documentation on what a protocol handler is. Perhaps like the following

```rust
/// A ProtocolHandler abstracts access to an underlying file/object storage.
/// It maps maps strings (e.g. URLs, filesystem paths, etc) to sources of bytes
```

Reply:
Thanks, I'll add this comment in the next commit

Reply:
BTW, `ProtocolHandler` or `ObjectStore` both sound fine to me, I can move to either name later. cc @houqp thoughts?